### PR TITLE
[13.0][IMP] account_bank_statement_import_txt_xlsx: header/footer trimming, separate debit/credit

### PR DESCRIPTION
**Pull request by zaoral**
Re open PR https://github.com/OCA/bank-statement-import/pull/455

---- 

Cherry pick of this PR for v12 https://github.com/OCA/bank-statement-import/pull/296
With some fixes and improvements

### CODE REVIEW

**Comment by zaoral** [comment]
Hi @alexey-pelykh,

I ask you for some help with this pr as see you are the maintainer. I have created this new PR as a reborn of what we have it #455 That PR was functionally tested and working fine we actually are usining in production: that pr was also reviewed by @rafaelbn but since we have some errors in the tests we were not able to continue with the merge.

In this new PR, I tried to fix the test but I am not able to reproduce locally the error we have from `TypeError: '_csv.reader' object is not subscriptable` more detail please review https://github.com/OCA/bank-statement-import/actions/runs/3440288290/jobs/5738578519 I want to reproduce locally actually review the error and fixed.

Instead, I am getting the next message for all the unit tests I run (see log below). It is like the statement is not even created, I think maybe something is missing o misconfigured in my environment. I did not have any other error or warning while running tests or updating the module. I also tried to use the same commands used in the GitHub action in my local environment but I haven't had luck.

Please if you know what I can do or check, to be able to reproduce the error locally in order to properly test and fix the pr please let me know.

Best Regards

```
2022-11-11 21:35:34,060 531 INFO bank-statement odoo.modules.registry: module account_bank_statement_import_txt_xlsx: creating or updating database tables 
2022-11-11 21:35:34,346 531 INFO bank-statement odoo.modules.loading: loading account_bank_statement_import_txt_xlsx/security/ir.model.access.csv 
2022-11-11 21:35:34,393 531 INFO bank-statement odoo.modules.loading: loading account_bank_statement_import_txt_xlsx/data/map_data.xml 
2022-11-11 21:35:34,402 531 INFO bank-statement odoo.modules.loading: loading account_bank_statement_import_txt_xlsx/views/account_bank_statement_import_sheet_mapping.xml 
2022-11-11 21:35:34,483 531 INFO bank-statement odoo.modules.loading: loading account_bank_statement_import_txt_xlsx/views/account_bank_statement_import.xml 
2022-11-11 21:35:34,552 531 INFO bank-statement odoo.modules.loading: loading account_bank_statement_import_txt_xlsx/views/account_journal_views.xml 
2022-11-11 21:35:34,665 531 INFO bank-statement odoo.modules.loading: loading account_bank_statement_import_txt_xlsx/wizards/account_bank_statement_import_sheet_mapping_wizard.xml 
2022-11-11 21:35:34,719 531 INFO bank-statement odoo.modules.loading: Module account_bank_statement_import_txt_xlsx: loading demo 
2022-11-11 21:35:34,756 531 INFO bank-statement odoo.modules.module: odoo.addons.account_bank_statement_import_txt_xlsx.tests.test_account_bank_statement_import_txt_xlsx running tests. 
2022-11-11 21:35:34,756 531 INFO bank-statement odoo.addons.account_bank_statement_import_txt_xlsx.tests.test_account_bank_statement_import_txt_xlsx: Starting TestAccountBankStatementImportTxtXlsx.test_balance ... 
2022-11-11 21:35:35,223 531 INFO bank-statement odoo.addons.account_bank_statement_import_txt_xlsx.tests.test_account_bank_statement_import_txt_xlsx: ====================================================================== 
2022-11-11 21:35:35,223 531 ERROR bank-statement odoo.addons.account_bank_statement_import_txt_xlsx.tests.test_account_bank_statement_import_txt_xlsx: FAIL: TestAccountBankStatementImportTxtXlsx.test_balance
Traceback (most recent call last):
  File "/home/odoo/custom/repositories/oca-bank-statement-import/account_bank_statement_import_txt_xlsx/tests/test_account_bank_statement_import_txt_xlsx.py", line 334, in test_balance
    self.assertEqual(len(statement), 1)
AssertionError: 0 != 1

```

**Comment by alexey-pelykh** [comment]
@zaoral your tests fail because the code makes no sense.

Before the change, whenever CSV was parsed, it iterated over rows as `for row in rows` without involving index. You're assuming that CSV file can be accessed as [sequence](https://docs.python.org/3/glossary.html#term-sequence) but it's only an [iterable](https://docs.python.org/3/glossary.html#term-iterable). Therefore you'd want either to read entire CSV to a list first (which is not optimal for huge files) or rework code further skip first `stat_first_index` rows and don't parse last `stat_last_index` of them but CSV won't give you the entire number of rows until you read them all.

Likely you can not reproduce it locally since you're working with XLSX files and not CSV ones..

**Comment by alexey-pelykh** [comment]
https://www.pythonmorsels.com/p/22p5u/

**Comment by zaoral** [comment]
Hi @alexey-pelykh 
Thank you for your response and help.

In this case, I am working on the changes to make it work.

NOTE: Also, the problem related to why I could not reproduce the error locally doing functional test neither running unit tests was because the environment I was using was auto-installing the enterprise module for CSV bank statement import, and the CSV functionally of this module was not actually used. 

Soon I will add more commits with the needed changes

Best Regards

**Comment by zaoral** [comment]
Hi @alexey-pelykh ! I have updated the pr in other that the code has a sense for CSV files, add new unit tests for both CSV and xlsx that use header/footer that need to be trimmed, and ensure that the new modifications of the code also will work for the pre-existing cases. I think the changes will be clearer to see in this link where only the new [forced changes](https://github.com/OCA/bank-statement-import/compare/2d04f7d0c6fba3869e9a3a547dd670911fc113a8..cffd85be778499174a9ecc24ae9c834130d2438c) are shown 

We look forward to adding this functionality to the oca, if you think that more changes need to be made please let me know

Thank you for your help in advance
Best Regards

**Comment by zaoral** [comment]
Hi @rafaelbn ! some months you made the review of https://github.com/OCA/bank-statement-import/pull/455, that pr was closed but this one was the continuation of it. If you also want to review this one with the last changes you are welcome.

Best Regards

File: account_bank_statement_import_txt_xlsx/models/account_bank_statement_import_sheet_parser.py [hunk 1]
```diff
@@ -51,9 +53,12 @@ def parse(self, data_file, mapping):
         if not lines:
             return currency_code, account_number, [{"transactions": []}]
 
-        lines = list(sorted(lines, key=lambda line: line["timestamp"]))
```
Comment (alexey-pelykh) [inline account_bank_statement_import_txt_xlsx/models/account_bank_statement_import_sheet_parser.py:64]: Why did you remove sorting? Do you anticipate that data in CSV is always ordered by date?
